pkg/registry/blobs.go

```diff
@@ -74,38 +94,96 @@ func (b *blobs) handle(resp http.ResponseWriter, req *http.Request) *regError {
 
 	switch req.Method {
 	case http.MethodHead:
-		b.lock.Lock()
-		defer b.lock.Unlock()
-		b, ok := b.contents[target]
-		if !ok {
+		h, err := v1.NewHash(target)
+		if err != nil {
+			return &regError{
+				Status:  http.StatusBadRequest,
+				Code:    "NAME_INVALID",
+				Message: err.Error(),
+			}
+		}
+		repo, err := name.NewRepository(req.URL.Host + path.Join(elem[1:len(elem)-2]...))
```

Comment:
(Resurrecting this because I'm sad we couldn't make it happen)

Aside from the question of whether we need to depend on `pkg/name` here, I wonder if `req.URL.Host` is adding anything besides letting the string pass through `NewRepository`. Really all blob storers want is the subrepo path.

I hear the concern about duplicating validation logic though. Would it be reasonable to export a method from `pkg/name` to validate a subrepo path just so `pkg/registry` could use it?

Reply:
> I'm sad we couldn't make it happen

Me too, but alas I've unblocked myself (and it is glorious 😉 ).


> I hear the concern about duplicating validation logic though

We have a bigger issue here.  This repo is now apparently governed by unwritten rules in one person's head, which also have had no rationale articulated anywhere that was arrived at by consensus of the stakeholders in this library.

I've clearly demonstrated a willingness to send PRs here (and elsewhere) to fix things, move things forward, and I'm happy to engage in productive discussions about them, but waiting a week to get a Nack based on unwritten rules (**that are already violated with what is checked in**), is frankly both aggravating and unsustainable.

Reply:
https://github.com/google/go-containerregistry/blob/a0c4bd256482b8522065d5f6cf966281ef270680/pkg/registry/registry.go#L15-L23

Reply:
In order for this to disqualify this change, "low dependency" must be interpreted to exclude *what it currently imports*.  🙃

Reply:
Also, I'd recommend a README: https://github.com/google/go-containerregistry#philosophy

Reply:
> We have a bigger issue here. This repo is now apparently governed by unwritten rules in one person's head, which also have had no rationale articulated anywhere that was arrived at by consensus of the stakeholders in this library.

Hoping to address this in #1166 

Taking the issue of dependency philosophy aside, I think I'm still convinced that the dependency on `pkg/name` is unnecessary, or at least having a full `name.Repository` in the interface is. If not for validation of the repository path, you don't actually need a full `name.Repository`, and could just get by with `Get(string path, h v1.Hash)`, and not add any new dependencies. Path validation itself is [fairly simple](https://github.com/google/go-containerregistry/blob/v0.6.0/pkg/name/check.go#L35), and could be moved to `internal/something` and shared by `pkg/name` and `pkg/registry`.

Reply:
You are assuming details of the blob layer by stating that it can only care about path.  Every single cloud vendor registry uses host to route storage requests, Amazon even gives each customer their own hostname per region.  I doubt this is how you wrote kontain.me, but…. 

Hopefully you get my point.

Reply:
Ah I see what you're saying. The same code depending on pkg/registry with some blob storage impl, running at `foo.registry.example` and `bar.registry.example`, might decide to store those blobs differently based on the host.

Can we document this expectation in the doc comments for `BlobHandler`? It wasn't immediately clear to me (because I'm a big dumb dummy) why the host was included, except to make path validation easier.

Reply:
> Can we document this expectation in the doc comments for BlobHandler

Absolutely.